feat: Added force flag to install-app command

ChillarAnand opened the pull request:
Added a `--force` flag for install-app command (can be used to reinstall apps while ignoring duplicates)

Closes https://github.com/frappe/frappe/issues/15690

Docs: https://frappeframework.com/docs/v13/user/en/bench/bench-commands/edit?wiki_page_patch=b2e65d6d84

#

TODO:
- [x] Test (just to check if things properly force install)

ankush commented:
This should've resolved duplicate module def issue?  https://github.com/frappe/frappe/pull/16604 


Anyway, force install flag might help in some other cases tho.

ankush commented on frappe/installer.py:
```diff
@@ -246,7 +246,11 @@ def install_app(name, verbose=False, set_as_patched=True):
-		add_module_defs(name)
+		try:
+			add_module_defs(name)
+		except frappe.exceptions.DuplicateEntryError as e:
```
btw use `ignore_if_duplicate` instead of catching exception (doesn't work on PG)

ChillarAnand committed:
feat: Added force flag to install-app command

phot0n commented on frappe/installer.py:
```diff
@@ -266,7 +266,7 @@ def install_app(name, verbose=False, set_as_patched=True):
 
 	if name != "frappe":
-		add_module_defs(name)
+		add_module_defs(name, ignore_if_duplicate=True)
```
```suggestion
		add_module_defs(name, ignore_if_duplicate=force)
```

need to add the same in the function definition 
https://github.com/frappe/frappe/blob/da32a2615e710b5c450ccf3b7dece160ee76763e/frappe/installer.py#L576

as well as 👇 
https://github.com/frappe/frappe/blob/da32a2615e710b5c450ccf3b7dece160ee76763e/frappe/installer.py#L582

phot0n commented on frappe/installer.py:
```diff
@@ -573,13 +573,13 @@ def make_site_dirs():
 
 
-def add_module_defs(app):
+def add_module_defs(app, ignore_if_duplicate):
```
```suggestion
def add_module_defs(app, ignore_if_duplicate=False):
```

phot0n reviewed (commented):
force argument needs to be passed in these places as well 👇 

https://github.com/frappe/frappe/blob/da32a2615e710b5c450ccf3b7dece160ee76763e/frappe/installer.py#L83

https://github.com/frappe/frappe/blob/da32a2615e710b5c450ccf3b7dece160ee76763e/frappe/installer.py#L246

phot0n commented:
@gavindsouza should we allow people to install apps if they're already installed 😬. 

basically:

```py
if not force and name in installed_apps:
   raise Exception("App {0} already installed".format(name))
```

as that is  what `--force` flag should mean (i think)

phot0n commented:
```diff
diff --git a/frappe/core/doctype/file/file.py b/frappe/core/doctype/file/file.py
index 3547a03832..8ca0b9ea10 100755
--- a/frappe/core/doctype/file/file.py
+++ b/frappe/core/doctype/file/file.py
@@ -608,7 +608,7 @@ def on_doctype_update():
 def make_home_folder():
        home = frappe.get_doc(
                {"doctype": "File", "is_folder": 1, "is_home_folder": 1, "file_name": _("Home")}
-       ).insert()
+       ).insert(ignore_if_duplicate=True)
 
        frappe.get_doc(
                {
@@ -618,7 +618,7 @@ def make_home_folder():
                        "is_attachments_folder": 1,
                        "file_name": _("Attachments"),
                }
-       ).insert()
+       ).insert(ignore_if_duplicate=True)
 
 
 @frappe.whitelist()
diff --git a/frappe/installer.py b/frappe/installer.py
index 356b1b3c9c..9a362ef630 100644
--- a/frappe/installer.py
+++ b/frappe/installer.py
@@ -251,9 +251,8 @@ def install_app(name, verbose=False, set_as_patched=True, force=False):
        if name not in frappe.get_all_apps():
                raise Exception("App not in apps.txt")
 
-       if name in installed_apps:
-               frappe.msgprint(frappe._("App {0} already installed").format(name))
-               return
+       if not force and name in installed_apps:
+               raise Exception("App {0} already installed".format(name))
 
        print("\nInstalling {0}...".format(name))
 
diff --git a/frappe/utils/install.py b/frappe/utils/install.py
index 08a399e5fe..4342fb037f 100644
--- a/frappe/utils/install.py
+++ b/frappe/utils/install.py
@@ -46,6 +46,7 @@ def after_install():
 
        if not frappe.conf.skip_setup_wizard:
                frappe.db.set_default("desktop:home_page", "setup-wizard")
+               frappe.db.set_value("System Settings", "System Settings", "setup_complete", 0)
 
        # clear test log
        with open(frappe.get_site_path(".test_log"), "w") as f:
```

from top to bottom:
- those ignore_if_duplicate flags are required there - as when _force_ reinstalling frappe the home folder is already present which causes duplicate key error
- the `not force` thing will help us force reinstall an app (and the msgprint doesnt really print anything on the console as we've not passed `raise_exception` argument - we could use `click.secho` ?)
- the set_value thing is required as when frappe force reinstalls and the setup had been completed in the previous install, the `setup_complete` value is already present in the db (and set to `1` - plus we set the `desktop:home_page` to be `setup-wizard` in the after install hook) which doesn't let us proceed further when we complete the setup after a reinstall and we're stuck on the setup wizard page :( - as seen [here](https://github.com/frappe/frappe/blob/ee40bda491c163a25e9341247b29c8602fd0850f/frappe/desk/page/setup_wizard/setup_wizard.py#L52-L53)

~~This makes me think, should the flag be really named `--force` ..we're essentially ignoring things and letting things somehow "_install_".~~
~~it should probably be called `--ignore-duplicates` (?) 🤔~~

phot0n commented on frappe/installer.py:
```diff
@@ -80,7 +80,7 @@ def _new_site(
 
 	for app in apps_to_install:
-		install_app(app, verbose=verbose, set_as_patched=not source_sql)
+		install_app(app, verbose=verbose, set_as_patched=not source_sql, force=force)
```
force flag is not really needed over here :P 
as at that stage we've created a fresh db - but should be fine to leave as it is (i think 😬 ).
(my bad)

phot0n reviewed (commented):
@ChillarAnand could you have a look at the above [comment](https://github.com/frappe/frappe/pull/16747#issuecomment-1114926576) and see if the changes make sense (and work fine for you)?
(was able to _force_ reinstall frappe as well with them)

can we do `force=force` here 👇 ?
https://github.com/frappe/frappe/blob/da32a2615e710b5c450ccf3b7dece160ee76763e/frappe/installer.py#L271

 the force flag there is used to skip the checking of `migration_hash` ...if the app is already installed, and if we forcefully reinstall it, it will get skipped and on the first/initial installation of the app that is anyway skipped.
(Hopefully it might work out :P)

we can remove the force flag from these places as we don't really use them there 👇 
https://github.com/frappe/frappe/blob/da32a2615e710b5c450ccf3b7dece160ee76763e/frappe/modules/import_file.py#L147

https://github.com/frappe/frappe/blob/da32a2615e710b5c450ccf3b7dece160ee76763e/frappe/modules/import_file.py#L206

phot0n committed:
chore: add comments and add force flag to more places

phot0n committed:
chore: use secho instead of raising exception if app is already installed

* chore: singular click import

phot0n committed:
fix: don't add navbar items if theyre already present

* fix: don't set installed_apps global value in after_install hook

phot0n commented on frappe/utils/install.py:
```diff
@@ -18,9 +18,6 @@ def before_install():
 
 
 def after_install():
-	# reset installed apps for re-install
-	frappe.db.set_global("installed_apps", '["frappe"]')
-
```
this is not needed here as we update `installed_apps` default key before running the `after_install` hook in the `install_apps` function (in `add_to_installed_apps` function)

https://github.com/frappe/frappe/blob/8a1889301a85d74fa88f9fbb8f6f8d4e42069b6b/frappe/installer.py#L273-L281

phot0n commented on frappe/utils/install.py:
```diff
@@ -241,6 +239,10 @@ def add_country_and_currency(name, country):
 def add_standard_navbar_items():
 	navbar_settings = frappe.get_single("Navbar Settings")
 
+	# don't add settings/help options if they're already present
+	if navbar_settings.settings_dropdown and navbar_settings.help_dropdown:
+		return
+
```
added this here because say we have erpnext/any other app installed which extends the navbar items and if we force install frappe then, we'll get a validation error as we check the number of navbar items before and after save.

skipping this function if the items are already present

phot0n commented on frappe/installer.py:
```diff
@@ -75,12 +77,18 @@ def _new_site(
 		no_mariadb_socket=no_mariadb_socket,
 	)
-	apps_to_install = (
+	apps_to_install = set(
```
using a set here because if someone passes a list of app names in `install_apps` param and it already had frappe then it will try to reinstall frappe.

phot0n committed:
fix: only set home_page in after_install, if the key doesn't have value

phot0n committed:
test: test for force install-app

Note: this is a very minimal test, as frappe_docs doesn;t really have any doctypes

phot0n committed:
Merge branch 'develop' into atom

gavindsouza commented:
Good work on this @phot0n 🎉
